atest/testdata/variables/variable_types.robot

```diff
@@ -0,0 +1,352 @@
+    Should be equal    ${b}    []    type=list
+    Should be equal    ${c}    ${2.0}
+    Should be equal    ${d}    ${3.4}
+
+Variable assignment: Invalid type for list in multiple variable assignment
+    [Documentation]    FAIL    Unrecognized type 'bad'.
+    ${a: int}    @{b: bad} =    Create List    9    8    7
+
+Variable assignment: type can not be set as variable
+    [Documentation]    FAIL    Unrecognized type '\${type}'.
+    VAR    ${type}    int
+    ${a: ${type}} =    Set variable    123
+
+Variable assignment: type syntax is not resolved from variable
+    VAR    ${type}    x: int
+    ${${type}} =    Set variable    12
+    Should be equal    ${x: int}    12
+
+Variable assignment: extended
+    [Documentation]    FAIL
+    ...    ValueError: Return value 'kala' cannot be converted to integer.
+    Should be equal    ${OBJ.name}    dude    type=str
+    ${OBJ.name: int} =    Set variable    42
+    Should be equal    ${OBJ.name}    ${42}    type=int
+    ${OBJ.name: int} =    Set variable    kala
+
+Variable assignment: item
+    [Documentation]    FAIL
+    ...    ValueError: Return value 'kala' cannot be converted to integer.
+    VAR    @{x}    1    2
+    ${x: int}[0] =    Set variable    3
+    Should be equal    ${x}    [3, "2"]    type=list
+    ${x: int}[0] =    Set variable    kala
+
+User keyword
+    Keyword    1    1    int
+    Keyword    1.2    1.2    float
+    Varargs    1    2    3
+    Kwargs    a=1    b=2.3
+    Combination of all args    1.0    2    3    4    a=5    b=6
+
+User keyword: default value
+    Default
+    Default    1
+    Default as string
+    Default as string    ${42}
+
+User keyword: wrong default value 1
+    [Documentation]    FAIL
+    ...    ValueError: Argument default value 'arg' got value 'wrong' that cannot be converted to integer.
+    Wrong default
+
+User keyword: wrong default value 2
+    [Documentation]    FAIL
+    ...    ValueError: Argument 'arg' got value 'yyy' that cannot be converted to integer.
+    Wrong default    yyy
+
+User keyword: invalid value
+    [Documentation]    FAIL
+    ...    ValueError: Argument 'type' got value 'bad' that cannot be \
+    ...    converted to 'int', 'float' or 'third value in literal'.
+    Keyword    1.2    1.2    bad
+
+User keyword: invalid type
+    [Documentation]    FAIL
+    ...    Invalid argument specification: \
+    ...    Invalid argument '\${arg: bad}': \
+    ...    Unrecognized type 'bad'.
+    Bad type
+
+User keyword: Invalid assignment with kwargs k_type=v_type declaration
+    [Documentation]    FAIL
+    ...    Invalid argument specification: \
+    ...    Invalid argument '\&{kwargs: int=float}': \
+    ...    Unrecognized type 'int=float'.
+    Kwargs does not support key=value type syntax
+
+Embedded arguments
+    [Tags]    kala
+    Embedded 1 and 2
+    Embedded type 1 and no type 2
+    Embedded type with custom regular expression 111
+    VAR    ${x}    1
+    VAR    ${y}    2
+    Embedded ${x} and ${y}
```

Comment:
Embedded args with variables would deserve its own test. More importantly, variable that's not a string should be tested.